src/util.h

```diff
@@ -24,16 +24,44 @@
 
 namespace node {
 
+// Hack around the compiler complaining about sizeof(void).
+template<typename T>
+struct size_of {
+  enum { value = sizeof(T) };
+};
+
+template<>
+struct size_of<void> {
+  enum { value = 1 };
+};
+
 // These should be used in our code as opposed to the native
 // versions as they abstract out some platform and or
 // compiler version specific functionality
 // malloc(0) and realloc(ptr, 0) have implementation-defined behavior in
 // that the standard allows them to either return a unique pointer or a
 // nullptr for zero-sized allocation requests.  Normalize by always using
 // a nullptr.
-inline void* Realloc(void* pointer, size_t size);
-inline void* Malloc(size_t size);
-inline void* Calloc(size_t n, size_t size);
+template<typename T>
+inline T* Realloc(T* pointer, size_t n, size_t size = size_of<T>::value);
+template<typename T>
+inline T* Malloc(size_t n, size_t size = size_of<T>::value);
+template<typename T>
+inline T* Calloc(size_t n, size_t size = size_of<T>::value);
+
+// Same things, but aborts immediately instead of returning nullptr when
+// no memory is available.
+template<typename T>
+inline T* CheckedRealloc(T* pointer, size_t n, size_t size = size_of<T>::value);
+template<typename T>
+inline T* CheckedMalloc(size_t n, size_t size = size_of<T>::value);
+template<typename T>
+inline T* CheckedCalloc(size_t n, size_t size = size_of<T>::value);
```

Comment:
I would like to make two suggestions:

1. Make checked alloc/realloc the default.  IOW, make unchecked opt-in instead of opt-out.  There is almost never any going forward when memory allocation fails.

2. For programmer economy, provide non-template versions that default to `T = char`.